### ISSUE TITLE
There are some serious, serious memory issues here - fixes provided

### ISSUE THREAD

**Issue by radekg** [mqttjs/MQTT.js#27]
Hi,

I have been working with a fork of your lib for quite a while now and I've managed to identify a number of them. First of all - the most obvious, undeclared variables. Here's a commit that fixes those:

https://github.com/radekg/MQTT.js/commit/5b7a1aa3fb581dde3cd1320bb2def89ebc1e1a3f
and one more here (in parse.js):
https://github.com/radekg/MQTT.js/commit/fddfaa33d832c0990223742475565a65fcbcb2e4

But the biggest memory issue is in `connection.js`, I was running 10000 connections for about 18 hours, each publishing 3.5KB of payload every 4 minutes and the memory was constantly growing, no matter what I would do to my code. I managed to nail it in finally.

The issue is that your main buffer in Connection object always grows. Always. Here's the fix:

https://github.com/radekg/MQTT.js/commit/f31ac1d9cdf66a27b7d748e79cfd23843465ca13
https://github.com/radekg/MQTT.js/commit/fec6b1bff81c353d88d9ffc3ace1186dc9279bc0

Basically, after:

```
this.buffer.read = pos;
this.buffer.written = len;
```

You, sir, need:

```
// Discard the old data in the buffer to free up the RAM
var tmpBuf = new Buffer( this.buffer.written - this.buffer.read );
this.buffer.copy( tmpBuf, 0, this.buffer.read, this.buffer.written );
this.buffer = null;
this.buffer = tmpBuf;
this.buffer.read = 0;
this.buffer.written = tmpBuf.length;
tmpBuf = null;
```

Also, when resizing the buffer at the top of `parse` there's not need to allocate more than new buffer.

In the last commit, in parse.js there's also a few memory management lines that will not hurt.
Hope this helps someone. Or just use the fork:

https://github.com/radekg/MQTT.js

And with npm:

```
npm install git@github.com:radekg/MQTT.js.git#master
```
